modules/sdk/src/main/scala/org/tessellation/sdk/http/p2p/middlewares/PeerAuthMiddleware.scala

```diff
@@ -0,0 +1,216 @@
+package org.tessellation.sdk.http.p2p.middleware
+
+import java.security.{PrivateKey, PublicKey}
+
+import cats.data.{Kleisli, OptionT}
+import cats.effect.{Async, Ref, Resource}
+import cats.syntax.applicative._
+import cats.syntax.applicativeError._
+import cats.syntax.contravariantSemigroupal._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.option._
+
+import org.tessellation.schema.cluster.TokenValid
+import org.tessellation.schema.peer.PeerId
+import org.tessellation.sdk.domain.cluster.services.Session
+import org.tessellation.sdk.domain.cluster.storage.SessionStorage
+import org.tessellation.sdk.http.p2p.headers.{`X-Id`, `X-Session-Token`}
+import org.tessellation.security.SecurityProvider
+import org.tessellation.security.signature.Signing
+
+import com.comcast.ip4s.{Host, Port}
+import fs2.{Chunk, Stream}
+import org.http4s.Status.Unauthorized
+import org.http4s._
+import org.http4s.client.Client
+import pl.abankowski.httpsigner.http4s._
+import pl.abankowski.httpsigner.signature.generic.{GenericGenerator, GenericVerifier}
+import pl.abankowski.httpsigner.signature.{Generator, Verifier}
+import pl.abankowski.httpsigner.{HttpCryptoConfig, SignatureValid}
+
+object PeerAuthMiddleware {
+
+  private def unauthorized[F[_]] = Response[F](status = Unauthorized)
```

Comment:
I wonder maybe we could attach some more descriptive message to the `Unauthorized`. I guess we could do it using `.withReason`. So we have more understanding where this unauthorized comes from and what's the reason.

Reply:
Especially to use it for the local verifiers.

Reply:
@TheMMaciek I can add a ticket to have a better internal error handling for p2p auth layer because it will require additional work for sure.

Reply:
done, CONST-414